DOC-3356: New feature documentation for `tinymceai`. (#3976)
tinymce/tinymce-docs: modules/ROOT/partials/configuration/tinymceai_options.adoc

Configuration options are grouped by the feature they configure. Options in the General section apply to the plugin as a whole or to multiple features.

[[options-general]]
== General options

These options apply to the plugin overall or to multiple AI features (Chat, Quick Actions, Review).

[TIP]
====

> For context toolbar use cases (a toolbar that appears when text is selected), use the xref:quickbars.adoc[Quick Toolbars] plugin and configure `+quickbars_selection_toolbar+` instead.

Reviewer: ```suggestion
For context toolbar use cases (a toolbar that appears when text is selected), use the xref:quickbars.adoc[Quick Toolbars] plugin and configure `+quickbars_selection_toolbar+` with the desired toolbar buttons.
```
"instead" felt weird

Reviewer: Placement under Options feels a bit weird too? Maybe better in some UI configuration section??

Author: Updated

====

[[content_id]]
=== `+content_id+`

> A unique identifier for the document or content being edited. When set, chat history is scoped to this ID, allowing conversations to be preserved and associated with the specific document across sessions. When not set, all conversations are shown regardless of document.

Reviewer: ```suggestion
A unique identifier for the document or content being edited. When set, the chat history shown for the user is scoped to this ID, allowing conversations to be preserved and associated with the specific document across sessions. When not set, all conversations for the user are shown regardless of document.
```
clarifying, otherwise this sounds scary

*Type:* `+String+`

*Default value:* `+undefined+`

.Example
[source,js]
----
tinymce.init({
  selector: 'textarea',
  plugins: 'tinymceai',
  toolbar: 'tinymceai-chat tinymceai-quickactions tinymceai-review',
  content_id: 'document-123',
  // Required for authentication
  tinymceai_token_provider: () => {
    return fetch('/api/token').then(r => r.json());
  }
});
----

[[tinymceai_token_provider]]
=== `+tinymceai_token_provider+`

A function that returns a Promise resolving to an object with a `+token+` property containing the signed JWT token for authenticating with the {pluginname} service.

The JWT payload must include these required claims:

* `+iat+`: Issued at time (provided by JWT libraries)
* `+exp+`: Expiration time (tokens cannot exceed 24 hours; 5-15 minutes recommended)
* `+aud+`: The {productname} API key
* `+auth+`: Authentication object (see separate authentication documentation)
* `+sub+`: Unique user ID

*Type:* `+Function+` (`+() => Promise<{ token: string }>+`)

*Default value:* `+undefined+`

> For JWT setup and required claims, see xref:tinymceai-jwt-authentication-intro.adoc[JWT Authentication].

Reviewer: ```suggestion
*Type:* `+Function+` (`+() => Promise<{ token: string }>+`)

*Default value:* `+undefined+`

The JWT payload must include these required claims:

* `+iat+`: Issued at time (provided by JWT libraries)
* `+exp+`: Expiration time (tokens cannot exceed 24 hours; 5-15 minutes recommended)
* `+aud+`: The {productname} API key
* `+auth+`: Authentication object
* `+sub+`: Unique user ID

For more information about JWT setup, required claims, and the authentication object, see xref:tinymceai-jwt-authentication-intro.adoc[JWT Authentication].
```
this order might make more sense, especially with the link to the auth docs being right after it being mentioned?

The token endpoint may return the JWT in different formats. Handle both:

* **JSON response**: Endpoint returns `+{ "token": "eyJ..." }+`. Use `+fetch(url).then(r => r.json())+`.

> * **Plain text response**: Endpoint returns the raw JWT string. Use `+fetch(url).then(r => r.text()).then(token => ({ token }))+`.

Reviewer: I don't know if this is necessary? we had some confusion with our endpoints but I don't think that'll be a normal problem? and/or this should be implied by the rest of the description. but if someone said we should include it, am happy to. 
suggestion in case you want the easy remove:

```suggestion
```

Reviewer: I maybe take that back after looking at the example below actually. Alternative suggestion:

```suggestion
The function must return the token value within an object with a `+token+` property. Ensure the response from the token endpoint is handled correctly, based on the response format:

* **JSON response**: Endpoint returns `+{ "token": "eyJ..." }+`. Use `+fetch(url).then(r => r.json())+`.
* **Plain text response**: Endpoint returns the raw JWT string. Use `+fetch(url).then(r => r.text()).then(token => ({ token }))+`.
```

Author: Updated in [this commit](https://github.com/tinymce/tinymce-docs/pull/3976/changes/22f13d6772b8597c2140d5a8015bda6eea9af822)

[source,js]
----
tinymce.init({
  selector: 'textarea',
  plugins: 'tinymceai',
  toolbar: 'tinymceai-chat tinymceai-quickactions tinymceai-review',
  tinymceai_token_provider: () => {
    return fetch('/api/token').then(r => r.text()).then(token => ({ token }));
  }
});
----

[[tinymceai_sidebar_type]]
=== `+tinymceai_sidebar_type+`

Controls how the AI sidebar is displayed. With `+static+`, the sidebar renders inside the editor. With `+floating+`, it renders in a separate container outside the editor and can be dragged on the page.

*Type:* `+String+`

*Possible Values:* `+'static'+`, `+'floating'+`

*Default value:* `+'static'+`

.Example
[source,js]
----
tinymce.init({
  selector: 'textarea',
  plugins: 'tinymceai',
  toolbar: 'tinymceai-chat tinymceai-quickactions tinymceai-review',
  tinymceai_sidebar_type: 'floating',
  // Required for authentication
  tinymceai_token_provider: () => {
    return fetch('/api/token').then(r => r.json());
  }
});
----

[NOTE]
====
Changing this property dynamically (after the editor has been initialized) is not supported and can result in unpredictable behavior.
====

[[tinymceai_default_model]]
=== `+tinymceai_default_model+`

> The default AI model to use when no model is explicitly selected.

Reviewer: ```suggestion
The default AI model to use when no model is explicitly selected by the user. If undefined, the AI service will select the best model for speed, quality, and cost.
```
figured it might sound like there's no default if this isn't set. stole the wording from the other docs.

*Default value:* `+undefined+`

.Example
[source,js]
----
tinymce.init({
  selector: 'textarea',
  plugins: 'tinymceai',
  toolbar: 'tinymceai-chat tinymceai-quickactions tinymceai-review',
  tinymceai_default_model: 'gemini-2-5-flash',
  // Required for authentication
  tinymceai_token_provider: () => {
    return fetch('/api/token').then(r => r.json());
  }
});
----

[[tinymceai_allow_model_selection]]
=== `+tinymceai_allow_model_selection+`

Whether users can select a different AI model from the chat interface.

*Type:* `+Boolean+`

*Possible Values:* `+true+`, `+false+`

*Default value:* `+true+`

.Example
[source,js]
----
tinymce.init({
  selector: 'textarea',
  plugins: 'tinymceai',
  toolbar: 'tinymceai-chat tinymceai-quickactions tinymceai-review',
  tinymceai_allow_model_selection: false,
  // Required for authentication
  tinymceai_token_provider: () => {
    return fetch('/api/token').then(r => r.json());
  }
});
----

[[options-chat]]
== Options for Chat

These options configure the AI Chat sidebar, where users have interactive conversations with the AI and can add external sources for context.

[[tinymceai_chat_fetch_sources]]
=== `+tinymceai_chat_fetch_sources+`

> A function that returns a Promise resolving to an array of additional context source groups. Each group has `+label+`, optional `+icon+`, and `+sources+` array. Each source has `+id+`, `+label+`, and `+type+` (`+'web-resource'+` or `+'file'+`).

Reviewer: ```suggestion
A function that returns a Promise resolving to an array of additional context source groups. Each group has `+label+`, optional `+icon+`, and `+sources+` array. Each source has `+id+`, `+label+`, and `+type+` (`+'web-resource'+` or `+'file'+`). A source's `+id+` is used to fetch its data via the `+tinymceai_chat_fetch_source+` option.

This data is used to populate the sources menu, with each source group being rendered as a submenu. Users can select these sources as additional context for chat conversations.
```

Reviewer: worth leading with the description of what it does, before launching the function description?

Reviewer: alternative suggestion:
```suggestion
Populates the sources menu with submenus of files and web resources. Users can select these sources as additional context for chat conversations. 

Takes a function that returns a Promise resolving to an array of additional context source groups. Each group has `+label+`, optional `+icon+`, and `+sources+` array. Each source has `+id+`, `+label+`, and `+type+` (`+'web-resource'+` or `+'file'+`). A source's `+id+` is used to fetch its content via the `+tinymceai_chat_fetch_source+` option.
```

Author: Updated in [this commit](https://github.com/tinymce/tinymce-docs/pull/3976/commits/fcef564b265ef79b4d62774f95510e60df91312e)

*Type:* `+Function+` (`+() => Promise<Array>+`)

*Possible Values:* For source `+type+` property: `+'web-resource'+`, `+'file'+`

*Default value:* `+() => Promise.resolve([])+`

.Example
[source,js]
----
tinymce.init({
  selector: 'textarea',
  plugins: 'tinymceai',
  toolbar: 'tinymceai-chat tinymceai-quickactions tinymceai-review',
  tinymceai_chat_fetch_sources: async () => [
    {
      label: 'My Documents',
      icon: 'folder',
      sources: [
        { id: 'doc-1', label: 'Document 1', type: 'file' },
        { id: 'url-1', label: 'Web Page', type: 'web-resource' }
      ]
    }

>   ],

Reviewer: ```suggestion
  ],
  tinymceai_chat_fetch_source: async (id) => {
    const res = await fetch(`/api/documents/\$\{id\}`);
    const blob = await res.blob();
    const filename = `\$\{id\}.pdf`;
    return { type: 'file', file: new File([blob], filename, { type: blob.type }) };
  },
```
worth having both in this example as well, to show you need both? ignore if you want to build on concepts or something

Author: Updated in [this commit](https://github.com/tinymce/tinymce-docs/pull/3976/commits/31979ffbb15b11f7d55c0e713ef60c3e4f1373c0)

  // Required for authentication
  tinymceai_token_provider: () => {
    return fetch('/api/token').then(r => r.json());
  }
});
----

[[tinymceai_chat_fetch_source]]
=== `+tinymceai_chat_fetch_source+`

> A function that fetches the content for an additional source by ID. Receives the source `+id+` and returns a Promise resolving to the source content (either `+{ type: 'file', file: File }+` or `+{ type: 'web-resource', url: string }+`).

Reviewer: ```suggestion
A function that fetches the content for an additional source by ID. Receives the source `+id+` and returns a Promise resolving to the source content (either `+{ type: 'file', file: File }+` or `+{ type: 'web-resource', url: string }+`). The content is passed to the AI agent as additional context for the chat conversation.
```

Reviewer: not sure we usually explain what an option does, but I figure it could help. possibly should lead with this, instead of the description of the function tbh

    { label: 'Docs', sources: [{ id: 'doc-1', label: 'Document 1', type: 'file' }] }
  ],
  tinymceai_chat_fetch_source: async (id) => {
    const res = await fetch(`/api/documents/\$\{id\}`);
    const blob = await res.blob();
    const filename = `\$\{id\}.pdf`;
    return { type: 'file', file: new File([blob], filename, { type: blob.type }) };
  },
  // Required for authentication
  tinymceai_token_provider: () => {
    return fetch('/api/token').then(r => r.json());
  }
});
----

[[tinymceai_chat_welcome_message]]
=== `+tinymceai_chat_welcome_message+`

Customises the welcome message displayed in the Chat sidebar when starting a new conversation.

*Type:* `+String+`

*Default value:* A default message introducing the AI assistant and its capabilities.

.Example
[source,js]
----
tinymce.init({
  selector: 'textarea',
  plugins: 'tinymceai',
  toolbar: 'tinymceai-chat tinymceai-quickactions tinymceai-review',
  tinymceai_chat_welcome_message: '<p>Welcome! How can I help you today?</p>',
  // Required for authentication
  tinymceai_token_provider: () => {
    return fetch('/api/token').then(r => r.json());
  }
});
----

[[options-quickactions]]
== Options for Quick Actions

> These options configure the Quick Actions menu, which provides one-click AI transformations such as improve writing, fix grammar, translate, and change tone.

Reviewer: ```suggestion
These options configure the Quick Actions menu, which provides one-click AI transformations such as writing and grammar improvements, translation, and tone changes.
```
"AI transformations such as improve writing" felt ironic. wouldn't if capitalised? but this seems easier.

[[tinymceai_quickactions_menu]]
=== `+tinymceai_quickactions_menu+`

> Array of control IDs that define the order of items in the Quick Actions menu.

Reviewer: ```suggestion
Array of control IDs that define the order of items in the Quick Actions menu. The default value includes all predefined items, including `+ai-quickactions-custom+` which adds a submenu of custom actions defined by `+tinymceai_quickactions_custom+`.
```
since we've already gotten a feedback question about how to include custom actions in the menu

*Type:* `+Array+` of `+String+`

*Default value:*
[source,js]
----
[
  'ai-quickactions-chat-prompts',
  'ai-quickactions-improve-writing',
  'ai-quickactions-continue-writing',
  'ai-quickactions-check-grammar',
  'ai-quickactions-change-length',
  'ai-quickactions-change-tone',
  'ai-quickactions-translate',
  'ai-quickactions-custom'
]
----

.Example
[source,js]
----
tinymce.init({
  selector: 'textarea',
  plugins: 'tinymceai',
  toolbar: 'tinymceai-chat tinymceai-quickactions tinymceai-review',
  tinymceai_quickactions_menu: [
    'ai-quickactions-improve-writing',
    'ai-quickactions-check-grammar',
    'ai-quickactions-custom'
  ],
  // Required for authentication
  tinymceai_token_provider: () => {
    return fetch('/api/token').then(r => r.json());
  }
});
----

[[tinymceai_quickactions_chat_prompts]]
=== `+tinymceai_quickactions_chat_prompts+`

> Array of control IDs for the Chat Commands submenu within Quick Actions.

Reviewer: ```suggestion
Array of control IDs for the Chat Commands submenu within the AI Quick Actions menu.
```
I'm being pedantic adding in "AI" but just want to be super clear

*Type:* `+Array+` of `+String+`

*Default value:* `+['ai-chat-explain', 'ai-chat-summarize', 'ai-chat-highlight-key-points']+`

.Example
[source,js]
----
tinymce.init({
  selector: 'textarea',
  plugins: 'tinymceai',
  toolbar: 'tinymceai-chat tinymceai-quickactions tinymceai-review',
  tinymceai_quickactions_chat_prompts: [
    'ai-chat-explain',
    'ai-chat-summarize',
    'ai-chat-highlight-key-points'
  ],
  // Required for authentication
  tinymceai_token_provider: () => {
    return fetch('/api/token').then(r => r.json());
  }
});
----

[[tinymceai_quickactions_change_tone_menu]]
=== `+tinymceai_quickactions_change_tone_menu+`

> Array of control IDs for the Change Tone submenu.

Reviewer: ```suggestion
Array of control IDs for the Change Tone submenu within the AI Quick Actions menu.
```

*Type:* `+Array+` of `+String+`

*Default value:*
[source,js]
----
[
  'ai-quickactions-tone-casual',
  'ai-quickactions-tone-direct',
  'ai-quickactions-tone-friendly',
  'ai-quickactions-tone-confident',
  'ai-quickactions-tone-professional'
]
----

.Example
[source,js]
----
tinymce.init({
  selector: 'textarea',
  plugins: 'tinymceai',
  toolbar: 'tinymceai-chat tinymceai-quickactions tinymceai-review',
  tinymceai_quickactions_change_tone_menu: [
    'ai-quickactions-tone-casual',
    'ai-quickactions-tone-professional'
  ],
  // Required for authentication
  tinymceai_token_provider: () => {
    return fetch('/api/token').then(r => r.json());
  }
});
----

[[tinymceai_languages]]
=== `+tinymceai_languages+`

> Array of language options for the Translate submenu within Quick Actions. Each item has `+title+` (displayed in the menu) and `+language+` (language code sent to the API, for example `+'english'+`, `+'chinese'+`).

Reviewer: ```suggestion
Array of language options for the Translate submenu within the AI Quick Actions menu. Each item has `+title+` (displayed in the menu) and `+language+` (language code sent to the API, for example `+'english'+`, `+'chinese'+`).
```

*Type:* `+Array+` of `+Object+`

*Default value:*
[source,js]
----
[
  { title: 'English', language: 'english' },
  { title: 'Chinese (Simplified)', language: 'chinese' },
  { title: 'Spanish', language: 'spanish' },
  { title: 'German', language: 'german' },
  { title: 'Japanese', language: 'japanese' },
  { title: 'Portuguese', language: 'portuguese' },
  { title: 'Korean', language: 'korean' },
  { title: 'Italian', language: 'italian' },
  { title: 'Russian', language: 'russian' }
]
----

.Example
[source,js]
----
tinymce.init({
  selector: 'textarea',
  plugins: 'tinymceai',
  toolbar: 'tinymceai-chat tinymceai-quickactions tinymceai-review',
  tinymceai_languages: [
    { title: 'English', language: 'english' },
    { title: 'French', language: 'french' },
    { title: 'German', language: 'german' }
  ],
  // Required for authentication
  tinymceai_token_provider: () => {
    return fetch('/api/token').then(r => r.json());
  }
});
----

[[tinymceai_quickactions_custom]]
=== `+tinymceai_quickactions_custom+`

> Array of custom command objects rendered in the Custom Commands menu item. Each item can be type `+action+` (quick action with immediate preview) or type `+chat+` (opens in chat).

Reviewer: ```suggestion
Array of custom actions rendered in the Custom submenu within the AI Quick Actions menu. Each item can be type `+action+` (quick action with immediate preview) or type `+chat+` (opens in chat).
```
we used to call them "commands" - dropped that so we don't get confused with our usual definition of commands

Reviewer: unless there's also a separate menu item for these, outside the quick actions menu? we've talked about it; I'm not sure

* `+title+`: Text shown in the menu and chat history

> * `+prompt+`: The actual prompt sent to the AI

Reviewer: ```suggestion
* `+prompt+`: The prompt sent to the AI
```

  tinymceai_quickactions_chat_prompts: [
    'ai-chat-explain',
    'ai-chat-summarize',
    'ai-chat-highlight-key-points'
  ],
  // Required for authentication
  tinymceai_token_provider: () => {
    return fetch('/api/token').then(r => r.json());
  }
});
----

[[tinymceai_quickactions_change_tone_menu]]
=== `+tinymceai_quickactions_change_tone_menu+`

Array of control IDs for the Change Tone submenu within the AI Quick Actions menu.

*Type:* `+Array+` of `+String+`

*Default value:*
[source,js]
----
[
  'ai-quickactions-tone-casual',
  'ai-quickactions-tone-direct',
  'ai-quickactions-tone-friendly',
  'ai-quickactions-tone-confident',
  'ai-quickactions-tone-professional'
]
----

.Example
[source,js]
----
tinymce.init({
  selector: 'textarea',
  plugins: 'tinymceai',
  toolbar: 'tinymceai-chat tinymceai-quickactions tinymceai-review',
  tinymceai_quickactions_change_tone_menu: [
    'ai-quickactions-tone-casual',
    'ai-quickactions-tone-professional'
  ],
  // Required for authentication
  tinymceai_token_provider: () => {
    return fetch('/api/token').then(r => r.json());
  }
});
----

[[tinymceai_languages]]
=== `+tinymceai_languages+`

> Array of language options for the Translate submenu within the AI Quick Actions menu. Each item has `+title+` (displayed in the menu) and `+language+` (language code sent to the API, for example `+'english'+`, `+'chinese'+`).

Reviewer 2: Maybe change language code to language name?

Author: @spocke Updated in [this commit](https://github.com/tinymce/tinymce-docs/pull/3976/commits/5b8d55b795223eb9d80fc722b0a10f068742091a)
